service/controller/v17/encrypter/kms/kms.go

```diff
@@ -44,72 +45,146 @@ func NewEncrypter(c *EncrypterConfig) (*Encrypter, error) {
+			Tags: awstags.NewKMS(tags),
+		}
+
+		out, err := ctlCtx.AWSClient.KMS.CreateKey(in)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		keyID = out.KeyMetadata.KeyId
+
+		e.logger.LogCtx(ctx, "level", "debug", "message", "created encryption key")
 	}
 
-	_, err = sc.AWSClient.KMS.EnableKeyRotation(&kms.EnableKeyRotationInput{
-		KeyId: key.KeyMetadata.KeyId,
-	})
-	if err != nil {
-		return microerror.Mask(err)
+	// This is importand key roation is enabled before creation alias.
+	// Otherwise it is not guaranteed it will be reconciled. Right now we
+	// *always* enable rotation before aliasing the key. So it is enough
+	// reconcile aliasing properly to make sure rotation is enabled.
+	{
+		e.logger.LogCtx(ctx, "level", "debug", "message", "enabling encryption key rotation")
+
+		in := &kms.EnableKeyRotationInput{
+			KeyId: keyID,
+		}
+
+		_, err = ctlCtx.AWSClient.KMS.EnableKeyRotation(in)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		e.logger.LogCtx(ctx, "level", "debug", "message", "enabled encryption key rotation")
 	}
 
-	_, err = sc.AWSClient.KMS.TagResource(&kms.TagResourceInput{
-		KeyId: key.KeyMetadata.KeyId,
-		Tags:  k.getKMSTags(customObject),
-	})
-	if err != nil {
-		return microerror.Mask(err)
+	{
+		e.logger.LogCtx(ctx, "level", "debug", "message", "creating encryption key alias")
+
+		clusterID := key.ClusterID(customObject)
+		keyAlias := aws.String(toAlias(clusterID))
+
+		in := &kms.CreateAliasInput{
+			AliasName:   keyAlias,
+			TargetKeyId: keyID,
+		}
+
+		_, err = ctlCtx.AWSClient.KMS.CreateAlias(in)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		e.logger.LogCtx(ctx, "level", "debug", "message", "created encryption key alias")
 	}
 
 	return nil
 }
 
-func (k *Encrypter) DeleteKey(ctx context.Context, keyAlias string) error {
-	sc, err := controllercontext.FromContext(ctx)
+func (e *Encrypter) EnsureDeletedEncryptionKey(ctx context.Context, customObject v1alpha1.AWSConfig) error {
+	ctlCtx, err := controllercontext.FromContext(ctx)
 	if err != nil {
 		return microerror.Mask(err)
 	}
 
-	// Get the KMS key ID using the key alias.
-	key, err := sc.AWSClient.KMS.DescribeKey(&kms.DescribeKeyInput{
-		KeyId: aws.String(keyAlias),
-	})
-	if err != nil {
-		return microerror.Mask(err)
-	}
+	var keyID *string
+	{
+		e.logger.LogCtx(ctx, "level", "debug", "message", "finding out encryption key")
 
-	// Delete the key alias.
-	_, err = sc.AWSClient.KMS.DeleteAlias(&kms.DeleteAliasInput{
-		AliasName: aws.String(keyAlias),
-	})
-	if err != nil {
-		return microerror.Mask(err)
+		// TODO we should search by tags here in case alias failed to create and cluster was deleted early. Issue: https://github.com/giantswarm/giantswarm/issues/4262.
+		out, err := e.describeKey(ctx, customObject)
+		if IsKeyNotFound(err) {
+			e.logger.LogCtx(ctx, "level", "debug", "message", "did not find encryption key")
+			e.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
```

Comment:
Is resource cancelation missing here?

Reply:
No it isn't. It's a "basic" resource. `return` is enough here.

Reply:
Ah, ok.